### PR TITLE
Don't reload the UI when the keyboard layout changes

### LINKED ISSUE #11522: Switching OS input method resets Terminal colors
### Windows Terminal version (or Windows build number)

10.0.22000.258

### Other Software

Bash (in WSL)

### Steps to reproduce

1. Inside a `xterm` terminal, override the terminal's colors using escape sequences such as:

    ```sh
    # Set "red" color to hex(ab4642)
    printf '\033]4;%d;rgb:%s\033\\' 1 'ab/46/42';
    # Set "background" color to hex(181818)
    printf '\033]%d;rgb:%s\033\\' 11 '18/18/18';
    ```

    The full theme I use can be applied using the following script https://github.com/fnune/base16-shell/blob/master/scripts/base16-default-dark.sh. This method allows me to conveniently switch between dark and light themes, regardless of the OS I'm currently using.

1. Assuming you have multiple input methods configured in Windown  (e.g. I switch between "US" and "US-International" regularly), press either Ctrl+Shift or Win+Space to switch between input methods.

    Switching input from the task bar using the mouse also triggers this bug.

### Expected Behavior

Switching input methods in Windows has no impact on the configuration of my open Terminal sessions.

### Actual Behavior

The terminal's colors get reset to the color scheme configured in Terminal for the current session.

Before switching input methods:
image

After switching input methods:
image

### PR DESCRIPTION
**Pull request by zadjii-msft**
Alrighty, this is a silly solution to the problem. Go read #11522.

What if, we just didn't reload the UI for a settings reload? 

Terminal refreshes the appearance of panes, profiles, tabs, etc, in `TerminalPage::_RefreshUIForSettingsReload`. That's only called in `TerminalPage::SetSettings`, when we specify that we _didn't_ need to reload the UI. Currently, we only _don't_ reload the UI on an initial load of the settings (that makes sense, there's no UI to "**re**load" yet.)

So what if we did all of `TerminalPage::SetSettings`, but without the UI reload? We update the keybindings, the command palette, the current language, but _not_ the actual content of the tabs. 

### before
11522-test-before

### after
11522-test-after

Closes #11522?

<hr>

A caveat from discussion: 

This might have some _weird_ irreproducible bugs in it. Consider
1. Open two terminal windows
2. Install a WSL distro
3. Switch the KB layout

Both these windows will reload the settings, find that the new settings (& dynamic profile) are different, and will try to write the settings to the disk. One will win? Maybe? Both will probably end up with a full UI refresh, for determining that the settings file changed. 

But is this fine anyways? We're pretty sure there's a backlog item for "reload the settings on a distro install" anyways. And the value add of this fix is probably worth more than the very rare new edge case where this still happens.

### CODE REVIEW

File: src/cascadia/TerminalApp/AppLogic.cpp [hunk 1]
```diff
@@ -1055,7 +1058,7 @@ namespace winrt::TerminalApp::implementation
 
     // Method Description:
     // - Reloads the settings from the settings.json file.
-    void AppLogic::_ReloadSettings()
+    void AppLogic::_ReloadSettings(const bool keybindingsOnly)
```
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/AppLogic.cpp:1061]: nits...
- update method description
- I'd rename this `needRefreshUI` or something similar just to make it more general

**Review by carlos-zamora — APPROVED** [review]
I mean, I don't mind this solution. If anything, I bet there's a few other scenarios we want to apply the exact same thing to.

**Comment by DHowett** [comment]
You know, one of the things that weirds me out about this is that we don't need to _reload_ the settings at all... just re-apply their effects.

Would it be possible to _only reprocess the key bindings_ from the original model objects? That would address the bulk of the issue _and_ make for a more targeted fix...

**Review by DHowett — CHANGES_REQUESTED** [review]
(Per discussion, not sure this is the right solution -- we may benefit from iterating further!)

**Comment by zadjii-msft** [comment]
> Would it be possible to _only reprocess the key bindings_ from the original model objects?

@lhecker might need to back me up on this - could we just call `ActionMap::_RefreshKeyBindingCaches()`, and have that reload the keybindings for the new kbd layout? I assumed that we couldn't do it trivially - I guess I assumed the layering effects would be lost by this point, and we'd need to re-parse the settings to get the right layering. 

Now that I'm re-reading `ActionMap::_PopulateKeyBindingMapWithStandardCommands`, that seems like it'd work without a whole reparse

**Comment by lhecker** [comment]
`_RefreshKeyBindingCaches()` is not the most important thing that needs to be reloaded when the keyboard layout changes. The `KeyChord` instances need to be refreshed, as they translate scan-codes to vkeys via `MapVirtualKeyW` and those are then used for the `Hash()` they return. In other words, yeah this is 100% doable, but it needs a bit more work than just calling `_RefreshKeyBindingCaches()`.

**Comment by zadjii-msft** [comment]
Hmm. Updating the `KeyChord`s seems... hard. The ctor will evaluate the SC->VK on construction, and store that the same as if we parsed the VK directly. So we can't just make a new `_KeyMap` map, take all the old `KeyChord`s and just re-evaluate them. 

Not sure what there is to do here. I can try to look if there's a way to build a fresh Settings object and just steal the actions map? Maybe that would work?

**Comment by zadjii-msft** [comment]
I could theoretically make changes to `AppLogic::ReloadSettings`:
* `AppLogic::_TryLoadSettings` returns out the new Settings rather than placing it in `_settings`.
* then, it would: 
  ```c++
  if (SUCCEEDED(_settingsLoadedResult) && keybindingsOnly) {
    _settings.CopyActionsFrom(newSettingsObject);
    _page->UpdateKeyboardSettings()
    return;
  }
  // the rest of the handling
  ```
  * `CopyActionsFrom(other)` does basically what it would sound like, just replace the `globals.ActionsMap` with the new one
  * `UpdateKeyboardSettings` would just `_UpdateCommandsForPalette(); CommandPalette().SetActionMap(_settings.ActionMap());` to update any labels, etc.

Does that seem like it's sensible? How egregious of a layering violation does that feel like?

**Comment by DHowett** [comment]
I think _this_ is the problem:

> The ctor will evaluate the SC->VK on construction

The thing consuming the settings model object here should be on the hook for this.

If we made that change, would it be easier to do "reevaluate all SC/VK" as a follow-on?

**Comment by DHowett** [comment]
(Feel free to veto me)

**Comment by zadjii-msft** [comment]
This is super stale. Like 1.5 years is too long, and there have been major rewrites since. We need to surely just start fresh

**Pull request closed without merging**